src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsAdornmentManager.cs

```diff
@@ -0,0 +1,225 @@
+
+        private TextFormattingRunProperties GetFormat(IClassificationType classificationType)
+        {
+            return _formatMap.GetTextProperties(classificationType);
+        }
+
+        /// <summary>
+        /// Get the spans located on each line so that it can only display the first one that appears on the line
+        /// </summary>
+        private void GetSpansOnEachLine(NormalizedSnapshotSpanCollection changedSpanCollection,
+            Dictionary<int, IMappingTagSpan<InlineDiagnosticsTag>> map)
+        {
+            var viewLines = TextView.TextViewLines;
+
+            foreach (var changedSpan in changedSpanCollection)
+            {
+                if (!viewLines.IntersectsBufferSpan(changedSpan))
+                {
+                    continue;
+                }
+
+                var tagSpans = TagAggregator.GetTags(changedSpan);
+                foreach (var tagMappingSpan in tagSpans)
+                {
+                    if (!ShouldDrawTag(changedSpan, tagMappingSpan, out var mappedPoint))
+                    {
+                        continue;
+                    }
+
+                    // mappedPoint is known to not be null here because it is checked in the ShouldDrawTag method call.
+                    var lineNum = mappedPoint!.Value.GetContainingLine().LineNumber;
+
+                    // If the line does not have an associated tagMappingSpan and changedSpan, then add the first one.
+                    if (!map.TryGetValue(lineNum, out var value))
+                    {
+                        map.Add(lineNum, tagMappingSpan);
+                    }
+
+                    // Draw the first instance of an error, if what is stored in the map at a specific line is
+                    // not an error, then replace it. Otherwise, just get the first warning on the line.
+                    else if (value is not null && value.Tag.ErrorType is not PredefinedErrorTypeNames.SyntaxError)
+                    {
+                        map[lineNum] = tagMappingSpan;
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Iterates through the mapping of line number to span and draws the diagnostic in the appropriate position on the screen,
+        /// as well as adding the tag to the adornment layer.
+        /// </summary>
+        protected override void AddAdornmentsToAdornmentLayer_CallOnlyOnUIThread(NormalizedSnapshotSpanCollection changedSpanCollection)
+        {
+            // this method should only run on UI thread as we do WPF here.
+            Contract.ThrowIfFalse(TextView.VisualElement.Dispatcher.CheckAccess());
+            if (changedSpanCollection.IsEmpty())
+            {
+                return;
+            }
+
+            var viewLines = TextView.TextViewLines;
+            using var _ = PooledDictionary<int, IMappingTagSpan<InlineDiagnosticsTag>>.GetInstance(out var map);
+            GetSpansOnEachLine(changedSpanCollection, map);
```

Comment:
nit.  generally we would name this 'Add' instead of 'Get' as you're filling a collection passed in.